### PR TITLE
Migrate lstsq from TH to ATen (CPU)

### LINKED ISSUE #24726: Migrate `lstsq` from the TH to Aten (CPU)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

cc @vishwakftw @SsnL

### PR DESCRIPTION
**Pull request by muthuArivoli**
Fixes #24726

### CODE REVIEW

**Comment by muthuArivoli** [comment]
Performance:
ATen
average calculation time is 0.00004920 (s).
average calculation time is 0.00019945 (s).
average calculation time is 0.00089689 (s).
average calculation time is 0.00499118 (s).
average calculation time is 0.03124119 (s).
average calculation time is 1.35202494 (s).
average calculation time is 17.55059886 (s).

TH
average calculation time is 0.00004178 (s).
average calculation time is 0.00017873 (s).
average calculation time is 0.00084260 (s).
average calculation time is 0.00462029 (s).
average calculation time is 0.02985182 (s).
average calculation time is 1.41796799 (s).
average calculation time is 18.05408704 (s).


```python
import torch
import time

torch.manual_seed(0)
torch.set_num_threads(1)

#warm up
for i in [10, 100, 1000]:
    for _ in range(100):
        n = torch.randint(1, i, (1,)).item()
        m = torch.randint(1, i, (1,)).item()
        k = torch.randint(1, i, (1,)).item()
        b = torch.randn(n, k)
        a = torch.randn(n, m)
        torch.lstsq(b, a)

dims = [1, 50, 100, 250, 500, 1000, 5000, 10000]
tests = [0, 1000, 1000, 1000, 1000, 1000, 10, 2]
for i in range(1, len(dims)):
    fwd_t = 0
    for _ in range(tests[i]):
        n = torch.randint(dims[i-1], dims[i], (1,)).item()
        m = torch.randint(dims[i-1], dims[i], (1,)).item()
        k = torch.randint(dims[i-1], dims[i], (1,)).item()
        b = torch.randn(n, k)
        a = torch.randn(n, m)
        t1 = time.time()
        torch.lstsq(b, a)
        t2 = time.time()
        fwd_t = fwd_t + (t2 -t1)
    fwd_avg = fwd_t / tests[i]
    print("average calculation time is %.8f (s)." % (fwd_avg))
```

**Comment by muthuArivoli** [comment]
@mruberry PTAL.

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -5630,14 +5630,14 @@
...
-    CPU: legacy::cpu::_th_gels_out
+    CPU: lstsq_out
     CUDA: legacy::cuda::_th_gels_out
 
 - func: lstsq(Tensor self, Tensor A) -> (Tensor solution, Tensor QR)
   use_c10_dispatcher: full
   variants: method, function
   dispatch:
-    CPU: legacy::cpu::_th_gels
+    CPU: lstsq
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:5640]: Are these explicit dispatches still necessary on CPU?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -1167,4 +1182,74 @@ Tensor& lu_solve_out(Tensor& result, const Tensor& self, const Tensor& LU_data,
   return result;
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ lstsq ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+template <typename scalar_t>
+static void apply_lstsq(Tensor& B, Tensor& A) {
+#ifndef USE_LAPACK
+  AT_ERROR("lstsq: LAPACK library not found in compilation");
```
Comment (mruberry) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1190]: TORCH_CHECK(False, ...

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 3]
```diff
@@ -1167,4 +1182,74 @@ Tensor& lu_solve_out(Tensor& result, const Tensor& self, const Tensor& LU_data,
...
+    apply_lstsq<scalar_t>(B_working, A_working);
+  });
+
+  return std::tuple<Tensor, Tensor>(B_working, A_working);
+}
+
+std::tuple<Tensor&,Tensor&> lstsq_out(Tensor& B_out, Tensor& A_out, const Tensor& B, const Tensor& A) {
+  Tensor A_tmp, B_tmp;
+  std::tie(B_tmp, A_tmp) = native::lstsq(B, A);
+  A_out.resize_as_(A_tmp).copy_(A_tmp);
```
Comment (mruberry) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1250]: Use `resize_output()` here and below

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 4]
```diff
@@ -1167,4 +1182,74 @@ Tensor& lu_solve_out(Tensor& result, const Tensor& self, const Tensor& LU_data,
...
+  B_out.resize_as_(B_tmp).copy_(B_tmp);
+  return std::tuple<Tensor&, Tensor&>(A_out, B_out);
```
Comment (mruberry) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1252]: This is interesting that the return value lstsq is (B, A) but here the out tensors are returned as (A, B). I think that's going to need a comment to clarify what's happening.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 5]
```diff
@@ -1167,4 +1182,74 @@ Tensor& lu_solve_out(Tensor& result, const Tensor& self, const Tensor& LU_data,
...
+  lwork = static_cast<int>(wkopt);
+  Tensor work_tensor = at::empty({lwork}, A.options().dtype());
+  auto work = work_tensor.data_ptr<scalar_t>();
+
+  lapackGels<scalar_t>('N', m, n, nrhs, A_data, lda, B_data, ldb, work, lwork, &info);
+#endif
+}
+
+std::tuple<Tensor, Tensor> lstsq(const Tensor& B, const Tensor& A) {
+  TORCH_CHECK(A.dim() == 2, "A should have 2 dimensions, but has ", A.dim());
```
Comment (mruberry) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1218]: "Expected two dimensions for A, but got ", A.dim()

Same fix for the B dim check below.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 6]
```diff
@@ -1167,4 +1182,74 @@ Tensor& lu_solve_out(Tensor& result, const Tensor& self, const Tensor& LU_data,
...
+  TORCH_CHECK(A.size(-1) != 0, "A should not be empty");
```
Comment (mruberry) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1219]: This seems like it should be a.numel == 0? Also, text should be "Expected non-empty A."

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 7]
```diff
@@ -1167,4 +1182,74 @@ Tensor& lu_solve_out(Tensor& result, const Tensor& self, const Tensor& LU_data,
...
+  TORCH_CHECK(B.dim() == 1 || B.dim() == 2, "B should have 1 or 2 "
+      "dimensions, but has ", B.dim());
+  TORCH_CHECK(B.size(-1) != 0, "B should not be empty");
+
+  TORCH_CHECK(A.size(0) == B.size(0), "Expected A and B to have same size "
+      "at dim 0, but A has ", A.size(0), " rows and B has ", B.size(0), " rows");
+
+  Tensor B_working = B.clone();
+  if(B_working.dim() == 1) {
```
Comment (mruberry) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1228]: `if(` -> `if (`

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 8]
```diff
@@ -1167,4 +1182,74 @@ Tensor& lu_solve_out(Tensor& result, const Tensor& self, const Tensor& LU_data,
...
+    B_working.unsqueeze_(1);
+  }
+  if (!B_working.is_contiguous()) {
```
Comment (mruberry) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1231]: I think this can be optimized by checking if B is contiguous and, if not, calling contiguous instead of clone to acquire B_working? Otherwise this will copy B and then copy B_working later, too.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 9]
```diff
@@ -1167,4 +1182,74 @@ Tensor& lu_solve_out(Tensor& result, const Tensor& self, const Tensor& LU_data,
...
+    B_working = B_working.contiguous();
+  }
+  if (A.size(0) < A.size(1)) {
```
Comment (mruberry) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1234]: Add a comment for this transform.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 10]
```diff
@@ -1167,4 +1182,74 @@ Tensor& lu_solve_out(Tensor& result, const Tensor& self, const Tensor& LU_data,
   return result;
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ lstsq ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+template <typename scalar_t>
```
Comment (mruberry) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1187]: Add a comment here describing the function's calling convention and briefly describing how it's implemented.

**Comment by mruberry** [comment]
Hey @muthuArivoli! Sorry this was languishing for awhile. Please us (or me) if you're not getting a response on a PR with 2-3 days. Sometimes we get buried under notifications and it can be easy for PRs to fall through the cracks.

Overall this looks good. I made a few minor comments about the code. Currently lstsq is tested in test_torch.py. Check out test_lstsq. As part of this port, would you (1) move the test to test_linalg.py and (2) review it and ensure it's testing all the edge cases that trigger runtime errors, the out variant, and that its computation is tested properly, too? Finally, while torch.lstsq != numpy.linalg.lstsq (which is why we want to implement torch.linalg.lstsq in the future), can we write a test that uses numpy.linalg.lstsq as a reference for the correctness of our torch.lstsq? They should both be solving the same least squares problem for some inputs, after all.

Let me know your thoughts about the test extension.
